File: 12-1.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"strings"
)

type Ship struct {
	pos complex128
	dir complex128
}

func (s *Ship) move(heading complex128, distance int) {
	s.pos += heading * complex(float64(distance), 0)
}

func (s *Ship) rotate(degrees int) {
	r := float64(degrees) * math.Pi / 180
	x := math.Cos(r)*real(s.dir) - math.Sin(r)*imag(s.dir)
	y := math.Sin(r)*real(s.dir) + math.Cos(r)*imag(s.dir)
	s.dir = complex(x, y)
}

func main() {
	bytes, err := ioutil.ReadFile("12.in")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	lines := strings.Split(strings.TrimSpace(string(bytes)), "\n")

	s := Ship{dir: 1 + 0i}
	for _, x := range lines {
		action := x[0]
		value, _ := strconv.Atoi(x[1:])
		switch action {
		case 'N':
			s.move(1i, value)
		case 'S':
			s.move(-1i, value)
		case 'E':
			s.move(1+0i, value)
		case 'W':
			s.move(-1+0i, value)
		case 'F':
			s.move(s.dir, value)
		case 'L':
			s.rotate(value)
		case 'R':
			s.rotate(-value)
		}
	}

	fmt.Println(math.Abs(real(s.pos)) + math.Abs(imag(s.pos)))
}

File: 12_1.py
import math

directions = {"N": 0 + 1j, "S": 0 - 1j, "E": 1 + 0j, "W": -1 + 0j}


class Ship:
    def __init__(self, pos, dir):
        self.pos = pos
        self.dir = dir

    def move(self, dir, value):
        if dir == "F":
            d = self.dir
        else:
            d = directions[dir]
        self.pos += d * value

    def rotate(self, angle):
        x = math.cos(angle) * self.dir.real - math.sin(angle) * self.dir.imag
        y = math.sin(angle) * self.dir.real + math.cos(angle) * self.dir.imag
        self.dir = complex(x, y)


if __name__ == "__main__":
    with open("12.in") as f:
        input = [x.strip() for x in f.readlines()]

    s = Ship(0 + 0j, 1 + 0j)
    for x in input:
        action = x[0]
        value = int(x[1:])
        if action in "NSEWF":
            s.move(action, value)
        elif action in "LR":
            if action == "R":
                value *= -1
            value *= math.pi / 180
            s.rotate(value)

    print(s.pos)
    print(abs(s.pos.real) + abs(s.pos.imag))
